jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassResolver.java

```
@@ -0,0 +1,136 @@
+   * @param projectProperties properties containing plugin information and help messages.
+   * @return the name of the main class to be used for the container entrypoint.
+   * @throws MainClassInferenceException if no valid main class is configured or discovered.
+   */
+  public static String resolveMainClass(
+      @Nullable String mainClass, ProjectProperties projectProperties)
+      throws MainClassInferenceException {
+    // If mainClass is null, try to find via projectProperties.
+    if (mainClass == null) {
+      mainClass = getMainClassFromJar(projectProperties);
+    }
+
+    // If mainClass is still null, try to search in class files.
+    if (mainClass == null) {
+      mainClass = findMainClassInClassFiles(projectProperties, true);
+
+    } else if (!BuildConfiguration.isValidJavaClass(mainClass)) {
+      // If mainClass found in projectProperties is not valid, try to search in class files, but
+      // don't error if not found in class files.
+      String mainClassInClassFiles = findMainClassInClassFiles(projectProperties, false);
+      if (mainClassInClassFiles != null) {
+        mainClass = mainClassInClassFiles;
+      }
+    }
+
+    Preconditions.checkNotNull(mainClass);
+    if (!BuildConfiguration.isValidJavaClass(mainClass)) {
+      projectProperties.getLogger().warn("'mainClass' is not a valid Java class : " + mainClass);
+    }
+
+    return mainClass;
+  }
+
+  @Nullable
+  private static String getMainClassFromJar(ProjectProperties projectProperties) {
+    projectProperties
+        .getLogger()
+        .info(
+            "Searching for main class... Add a 'mainClass' configuration to '"
+                + projectProperties.getPluginName()
+                + "' to improve build speed.");
+    return projectProperties.getMainClassFromJar();
+  }
+
+  @Nullable
+  private static String findMainClassInClassFiles(
+      ProjectProperties projectProperties, boolean shouldError) throws MainClassInferenceException {
```

Making the method throw or suppress exceptions based on an argument seems weird to me. I'm thinking it's better to remove `shouldError` and have caller to catch the exception if they really want to ignore it.